File: src/services/street-view.ts
import type { AppConfig } from "../config.js";
import { hasUsableAddress, hasUsableCoordinates } from "../domain/address.js";
import type { Lot } from "../domain/types.js";

const STREET_VIEW_URL = "https://maps.googleapis.com/maps/api/streetview";

export class StreetViewClient {
  constructor(private readonly config: Pick<AppConfig, "googleApiKey" | "streetviewPitch" | "streetviewZoom" | "streetviewRadiusMeters" | "httpTimeoutMs">) {}

  buildUrl(lot: Lot): URL {
    if (this.config.googleApiKey === undefined) throw new Error("GOOGLE_API_KEY is required");
    const url = new URL(STREET_VIEW_URL);
    if (hasUsableAddress(lot.address)) {
      // This deliberately preserves the production Python request for addressed parcels.
      url.searchParams.set("location", `${lot.address}, CHICAGO, IL`);
    } else {
      if (!hasUsableCoordinates(lot)) throw new Error(`Lot ${lot.id} has neither a usable address nor coordinates`);
      url.searchParams.set("location", `${lot.lat},${lot.lon}`);
      url.searchParams.set("radius", String(this.config.streetviewRadiusMeters));
      url.searchParams.set("source", "outdoor");
    }
    url.searchParams.set("key", this.config.googleApiKey);
    url.searchParams.set("size", "1000x1000");
    url.searchParams.set("fov", "65");
    url.searchParams.set("pitch", String(this.config.streetviewPitch));
    url.searchParams.set("zoom", String(this.config.streetviewZoom));
    url.searchParams.set("return_error_code", "true");
    return url;
  }

  async fetchImage(lot: Lot): Promise<Uint8Array> {
    const url = this.buildUrl(lot);
    let lastError: unknown;
    for (let attempt = 1; attempt <= 3; attempt += 1) {
      try {
        const response = await fetch(url, { signal: AbortSignal.timeout(this.config.httpTimeoutMs) });
        if (!response.ok) throw new Error(`Street View returned HTTP ${response.status}`);
        const contentType = response.headers.get("content-type")?.toLowerCase() ?? "";
        if (!contentType.startsWith("image/")) throw new Error(`Street View returned unexpected content type: ${contentType || "missing"}`);
        const bytes = new Uint8Array(await response.arrayBuffer());
        if (bytes.byteLength === 0) throw new Error("Street View returned an empty image");
        return bytes;
      } catch (error) {
        lastError = error;
        if (attempt < 3) await new Promise((resolve) => setTimeout(resolve, 250 * 2 ** (attempt - 1)));
      }
    }
    throw lastError;
  }
}

File: src/streetview.rs
use crate::{
    config::Config,
    domain::{Lot, coordinates, usable_address},
    http::{Http, body},
};
use anyhow::{Result, bail, ensure};
use std::time::Duration;
use url::Url;
pub fn build_url(config: &Config, lot: &Lot) -> Result<Url> {
    let mut url = Url::parse("https://maps.googleapis.com/maps/api/streetview")?;
    let location = if usable_address(&lot.address) {
        let trimmed = lot.address.trim();
        let parts: Vec<_> = trimmed.split(',').map(str::trim).collect();
        let chicago =
            parts.len() >= 3 && parts[parts.len() - 2].eq_ignore_ascii_case("CHICAGO") && {
                let state = parts[parts.len() - 1].to_uppercase();
                state == "IL" || state.starts_with("IL ")
            };
        if chicago {
            trimmed.to_owned()
        } else {
            format!("{trimmed}, CHICAGO, IL")
        }
    } else {
        ensure!(
            coordinates(lot.lat, lot.lon),
            "Parcel has neither address nor coordinates"
        );
        format!("{},{}", lot.lat, lot.lon)
    };
    url.query_pairs_mut()
        .append_pair("location", &location)
        .append_pair("radius", &config.radius.to_string())
        .append_pair("source", "outdoor")
        .append_pair("key", config.required("GOOGLE_API_KEY")?)
        .append_pair("size", "1000x1000")
        .append_pair("fov", "65")
        .append_pair("pitch", &config.pitch.to_string())
        .append_pair("zoom", &config.zoom.to_string())
        .append_pair("return_error_code", "true");
    Ok(url)
}
pub fn fetch(http: &Http, url: &str) -> Result<Vec<u8>> {
    for attempt in 0..3 {
        let result = (|| {
            let response = http.get(url, None)?;
            ensure!(
                response.status().is_success(),
                "Street View returned HTTP {}",
                response.status().as_u16()
            );
            let content_type = response
                .headers()
                .get("content-type")
                .and_then(|v| v.to_str().ok())
                .unwrap_or_default();
            ensure!(
                content_type.to_ascii_lowercase().starts_with("image/"),
                "Street View returned an unexpected content type"
            );
            let bytes = body(response, 2_000_000)?;
            ensure!(!bytes.is_empty(), "Street View returned an empty image");
            Ok(bytes)
        })();
        match result {
            Ok(bytes) => return Ok(bytes),
            Err(error) if attempt == 2 => return Err(error),
            Err(_) => std::thread::sleep(Duration::from_millis(250 << attempt)),
        }
    }
    bail!("Street View request failed")
}
